src/DataLayouts/non_extruded_broadcasted.jl

```diff
@@ -0,0 +1,139 @@
+        f::F,
+        args::Tuple,
+        axes = nothing,
+    ) where {F}
+        # using Core.Typeof rather than F preserves inferrability when f is a type
+        return new{typeof(style), typeof(axes), Core.Typeof(f), typeof(args)}(
+            style,
+            f,
+            args,
+            axes,
+        )
+    end
+    function NonExtrudedBroadcasted(f::F, args::Tuple, axes = nothing) where {F}
+        NonExtrudedBroadcasted(combine_styles(args...)::BroadcastStyle, f, args, axes)
+    end
+    function NonExtrudedBroadcasted{Style}(f::F, args, axes = nothing) where {Style, F}
+        return new{Style, typeof(axes), Core.Typeof(f), typeof(args)}(
+            Style()::Style,
+            f,
+            args,
+            axes,
+        )
+    end
+    function NonExtrudedBroadcasted{Style, Axes, F, Args}(
+        f,
+        args,
+        axes,
+    ) where {Style, Axes, F, Args}
+        return new{Style, Axes, F, Args}(Style()::Style, f, args, axes)
+    end
+end
+
+@inline to_broadcasted(bc::NonExtrudedBroadcasted) =
+    Base.Broadcast.Broadcasted(bc.style, bc.f, bc.args, bc.axes)
+@inline to_non_extruded_broadcasted(bc::Base.Broadcast.Broadcasted) =
+    NonExtrudedBroadcasted(bc.style, bc.f, to_non_extruded_broadcasted_args(bc.args), bc.axes)
+@inline to_non_extruded_broadcasted(x) = x
+# NonExtrudedBroadcasted(bc::Base.Broadcast.Broadcasted) = to_non_extruded_broadcasted(bc)
+
+@inline to_non_extruded_broadcasted_args(args::Tuple) = (
+    to_non_extruded_broadcasted(args[1]),
+    to_non_extruded_broadcasted_args(Base.tail(args))...,
+)
+@inline to_non_extruded_broadcasted_args(args::Tuple{Any}) =
+    (to_non_extruded_broadcasted(args[1]),)
+@inline to_non_extruded_broadcasted_args(args::Tuple{}) = ()
+
+@inline _checkbounds(bc, _, I) = nothing # TODO: fix this case
+@inline _checkbounds(bc, ::Tuple, I) = Base.checkbounds(bc, I)
+@inline function Base.getindex(
+    bc::NonExtrudedBroadcasted,
+    I::Union{Integer, CartesianIndex},
+)
+    @boundscheck _checkbounds(bc, axes(bc), I) # is this really the only issue?
+    @inbounds _broadcast_getindex(bc, I)
+end
+
+# --- here, we define our own bounds checks
+@inline function Base.checkbounds(bc::NonExtrudedBroadcasted, I::Integer)
+    # Base.checkbounds_indices(Bool, axes(bc), (I,)) || Base.throw_boundserror(bc, (I,)) # from Base
+    Base.checkbounds_indices(Bool, (Base.OneTo(n_dofs(bc)),), (I,)) || Base.throw_boundserror(bc, (I,))
+end
+
+# To handle scalar cases, let's just switch back to
+# Base.Broadcast.Broadcasted and allow cartesian indexing:
+Base.@propagate_inbounds Base.getindex(bc::NonExtrudedBroadcasted) = to_broadcasted(bc)[CartesianIndex(())]
+
+import StaticArrays
+to_tuple(::Tuple{}) = ()
+to_tuple(t::Tuple) = t
+to_tuple(t::NTuple{N, <: Base.OneTo}) where {N} = map(x->x.stop, t)
+to_tuple(t::NTuple{N, <: StaticArrays.SOneTo}) where {N} = map(x->x.stop, t)
+n_dofs(bc::NonExtrudedBroadcasted) = prod(to_tuple(axes(bc)))
```

Comment:
```suggestion
n_dofs(bc::NonExtrudedBroadcasted) = prod(length, axes(bc))
```
Does this work? It's a little simpler to read.

Reply:
I can try

Reply:
Hm, I _think_ that should be fine, but it does assume that `axes` of `bc` is a `Base.OneTo` or `StaticArrays.SOneTo`. If it returned a tuple of `Int`s, then `length(4) # 1` will not error but give a wrong number for `n_dofs`. Ok, yeah, it looks like `axes` is always supposed to return a range, I can change it.

Reply:
It turns out that this didn’t work for some reason, so I reverted it.

Reply:
Hmm, that's strange. What's the example it didn't work for, and are we sure it's correct as is? I would expect the total number of degrees of freedom to be the product of the number of degrees of freedom along each axis, where the latter should correspond to the number of elements along the axis.

Reply:
Here is the build: https://buildkite.com/clima/climacore-ci/builds/4612, looks like a lot of jobs broke, actually

Reply:
Ah, I got the edge case wrong, I had `n_dofs(bc::NonExtrudedBroadcasted) = prod(length, axes(bc); init=0)` and it should apparently be `n_dofs(bc::NonExtrudedBroadcasted) = prod(length, axes(bc); init=1)`-- I didn't expect this behavior from `prod`:

```julia
julia> prod(())
1
```

Back to your question of is it correct as is: I think it's mostly correct, but I'll add some unit tests to give us more confidence.

Reply:
I think I fixed the remaining cases that I could think of.